Updated containers to use ABC

nwu63 opened the pull request:
## Purpose
I recently discovered abstract base classes in Python, and they seem like the correct approach for implementing these containers, since it's very clear which methods need to be implemented and which you get for free via inheritance. See for example [this article](https://treyhunner.com/2019/04/why-you-shouldnt-inherit-from-list-and-dict-in-python/) for a relevant discussion. Overall I think this is a much better implementation than what we had before, i.e. directly inheriting `dict` and `set`. A few things to note:
- I added a few new tests
- I added type annotations
- I clarified in docstrings that the initial capitalization is _persistent_, meaning `update()` etc. with a different capitalization will not replace it. I think this is the behaviour we want in relation to options dictionaries.

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [x] Maintenance update
- [ ] Other (please describe)

## Testing
I added a few more tests.

## Checklist
_Put an `x` in the boxes that apply._

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation

eirikurj commented on tests/test_CaseInsensitve.py:
```diff
@@ -93,45 +105,69 @@ def test_equal(self):
+            res.append((k, v))
+        self.assertEqual(res, [("opTION1", value2), ("optioN2", value2)])
+
+    def test_iter(self):
```
I suggest we add a test also using the `keys()` and `values()` methods, since that is somewhat used.

nwu63 replied:
Yep agreed, I have added those.

eirikurj reviewed (changes requested):
Looks like there is some regression in function. 
For instance, printing the object is not possible now, and `keys()` and `values()` do not return what you expect.

```
>>> from baseclasses.utils import CaseInsensitiveDict
>>> dd = CaseInsensitiveDict({"Aa": 22, "bb": 11})
>>> dd
<baseclasses.utils.CaseInsensitiveDict object at 0x7fed5fa0b880>
>>> dd.keys()
KeysView(<baseclasses.utils.CaseInsensitiveDict object at 0x7fed5fa0b880>)
>>> dd.values()
ValuesView(<baseclasses.utils.CaseInsensitiveDict object at 0x7fed5fa0b880>
```

To fix the printing and get similar behavior as before e.g., for the dict, we could add something like to the class
```
    def __repr__(self):
        return str(self.data)
```

This fixes the print, but the other methods are still broken,
```
>>> dd = CaseInsensitiveDict({"Aa": 22, "bb": 11})
>>> dd
{'Aa': 22, 'bb': 11}
>>> dd.keys()
KeysView({'Aa': 22, 'bb': 11})
>>> dd.values()
ValuesView({'Aa': 22, 'bb': 11})
```

nwu63 commented on baseclasses/utils.py:
```diff
@@ -1,31 +1,39 @@
-        self.map = {k.lower(): k for k in self.keys()}
+        self.data: dict = dict(*args, **kwargs)
+        if not all([isinstance(i, str) for i in self.data]):
+            raise ValueError("All keys must be strings!")
```
Good point.

nwu63 commented:
I pushed some updates that addressed your comments @eirikurj. Regarding `keys()` and `items()`, these behave as expected via the implemented `__iter__()` method, but as you pointed out the `__repr__` of these view objects do not match the default `dict` behaviour, i.e. printing out what's in the container. This is an inconsistency with how Python implemented the ABC and I don't think it's something we can change unless we subclass and extend these View classes ourselves and modify the `__repr__`. IMO this is not a big deal since in general you will want to either directly access the iterator `for k in d.keys()`, or explicitly cast to list via `list(d.keys())`.

nwu63 commented on baseclasses/utils.py:
```diff
@@ -1,31 +1,39 @@
+        self.map: Dict[str, str] = {k.lower(): k for k in self.data.keys()}
 
-    def _getKey(self, key):
+    def _getKey(self, key: str, raiseError=False):
```
Fixed

marcomangano reviewed (approved):
Nice stuff
